modules/ServerAPI/src/server/routes/FileServices.coffee

```diff
@@ -3,13 +3,229 @@ fs = require 'fs'
+		exists = await fs.promises.access(fullPath).then(() => true).catch(() => false)
+
+		if !exists
+			return []
+
+		# List all files in the folder recursively but exclude folders
+		files = await fs.promises.readdir fullPath, {withFileTypes: true}
+		filePaths = await Promise.all files.map((file) =>
+			filePath = path.join folderPath, file.name
+			if file.isDirectory()
+				if recursive
+					return await @listFiles filePath
+				else
+					# If recursive is false then return an empty array which will be flattened later
+					return []
+			else
+				return path: filePath
+		)
+		return Array.prototype.concat(...filePaths)
+
+class ExternalFileHandler extends LocalFileHandler
+
+class GCSFileHandler extends ExternalFileHandler
+
+	init: ->
+		@bucket = await @_getOrCreateBucket()
+
+	_getOrCreateBucket: ->
+		projectID = config.all.server.datafiles.gcs.projectID
+		bucketName = config.all.server.datafiles.gcs.bucketName
+		region = config.all.server.datafiles.gcs.bucketRegion
+		clientEmail = config.all.server.datafiles.gcs.clientEmail
+		privateKey = config.all.server.datafiles.gcs.privateKey
+		storage = new Storage(
+			projectId: projectID,
+			credentials: {
+				client_email: clientEmail,
+				private_key: privateKey
+			}
+		)
+		uniformBucketLevelAccess = true
+		defaultAcl = [{entity: 'allUsers',role: storage.acl.READER_ROLE}]
+		
+		bucket = await storage.bucket(bucketName)
+		[exists] = await bucket.exists()
+		if exists
+			console.log "Bucket #{bucketName} already exists."
+			[bucket] = await bucket.get()
+			return bucket
+		else
+			[bucket] = await storage.createBucket bucketName,
+				location: region
+				uniformBucketLevelAccess: uniformBucketLevelAccess
+				defaultAcl: defaultAcl
+				console.log "Bucket #{bucketName} created with uniform ACLs."
+			return bucket
+	
+	copyFile: (sourceLocation, targetLocation, metaData)->
+		if metaData == undefined
+			metaData = {}
+		else
+			metadata = {
+				metadata: metaData
+			};
+		[file] = await @bucket.upload(sourceLocation, {destination: targetLocation, metadata: metadata})
+		console.log "Uploaded file #{sourceLocation} to #{targetLocation} in bucket #{@bucket.name} with metadata #{JSON.stringify(metadata)}"
+		# Return both the file and the original file path
+		return {sourceLocation: sourceLocation, targetLocation: targetLocation, metaData: metaData}
+
+	getFile: (filePath) ->
+		file = await @bucket.file(filePath)
+		[exists] = await file.exists()
+		if !exists
+			throw new Error("File does not exist: #{filePath}")
+		metaData = await file.getMetadata()
+		fileStream = await file.createReadStream()
+		metaData[0].contentType = metaData[0].contentType || 'application/octet-stream'
+		return {fileStream: fileStream, metaData: metaData[0]}
+
+	listFiles: (folderName, recursive) ->
+		# If recursive is true, list all files in the folder and subfolders
+
+		if folderName == undefined
+			folderName = ''
+		if recursive == undefined
+			recursive = false
```

Comment:
It looks like recursive isn't used in the GCS handler. Is that intentional? Is the GCS bucket `getFiles` already recursive since it's just based on a prefix?